ImportError does not result in collection error

This appears to have been introduced in 0.23.2  (0.21.x works):

1. `pip3 install pytest pytest-asyncio`
2. Define the following file (or test)

`test_one.py`
```Python
import asdf # nonexistent module

def test_abc():
  pass
```

3. `pytest --import-mode=importlib` will report no tests run under `pytest-asyncio==0.23.2`, but will show the ImportError under `pytest-asyncio==0.21.1`. Note that the default import mode (`--import-mode=prepend`)  does not exhibit this problem.


Output when running under 0.23.2:
```
(.venv) williamjen@xps-32:~/Documents/import-error-pytest$ cat test_one.py 
import asdf

def test_abc():
    pass
(.venv) williamjen@xps-32:~/Documents/import-error-pytest$ pytest --import-mode=importlib
============================================================================================== test session starts ==============================================================================================
platform linux -- Python 3.10.12, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/williamjen/Documents/import-error-pytest
plugins: asyncio-0.23.2
asyncio: mode=strict
collected 0 items                                                                                                                                                                                               

============================================================================================= no tests ran in 0.01s =============================================================================================
(.venv) williamjen@xps-32:~/Documents/import-error-pytest$
```

Output when running under 0.21.1:
```
(.venv) williamjen@xps-32:~/Documents/import-error-pytest$ pytest --import-mode=importlib
============================================================================================== test session starts ==============================================================================================
platform linux -- Python 3.10.12, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/williamjen/Documents/import-error-pytest
plugins: asyncio-0.21.1
asyncio: mode=strict
collected 0 items / 1 error                                                                                                                                                                                     

==================================================================================================== ERRORS =====================================================================================================
_________________________________________________________________________________________ ERROR collecting test_one.py __________________________________________________________________________________________
ImportError while importing test module '/home/williamjen/Documents/import-error-pytest/test_one.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
test_one.py:1: in <module>
    import asdf
E   ModuleNotFoundError: No module named 'asdf'
============================================================================================ short test summary info ============================================================================================
ERROR test_one.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================
```

My suspicion is the `CollectError` caught in a214c3e77149608d427ccab69140edb509c67697 since the ImportError appears to be propagated via a `CollectError`, but I'm not sure.

--- a/docs/source/reference/changelog.rst
+++ b/docs/source/reference/changelog.rst
@@ -4,7 +4,7 @@ Changelog
 
 0.23.3 (UNRELEASED)
 ===================
-- Fixes a bug that caused event loops to be closed prematurely when using async generator fixtures with class scope or wider in a function-scoped test `#708 <https://github.com/pytest-dev/pytest-asyncio/issues/708>`_
+- Fixes a bug that caused event loops to be closed prematurely when using async generator fixtures with class scope or wider in a function-scoped test `#706 <https://github.com/pytest-dev/pytest-asyncio/issues/706>`_
 - Fixes a bug that caused an internal pytest error when using unittest.SkipTest in a module `#711 <https://github.com/pytest-dev/pytest-asyncio/issues/711>`_
 - Fixes a bug that caused an internal pytest error when an ImportWarning is emitted in a module `#713 <https://github.com/pytest-dev/pytest-asyncio/issues/713>`_
 

--- a/docs/source/reference/changelog.rst
+++ b/docs/source/reference/changelog.rst
@@ -5,8 +5,8 @@ Changelog
 0.23.3 (UNRELEASED)
 ===================
 - Fixes a bug that caused event loops to be closed prematurely when using async generator fixtures with class scope or wider in a function-scoped test `#706 <https://github.com/pytest-dev/pytest-asyncio/issues/706>`_
-- Fixes a bug that caused an internal pytest error when using unittest.SkipTest in a module `#711 <https://github.com/pytest-dev/pytest-asyncio/issues/711>`_
-- Fixes a bug that caused an internal pytest error when an ImportWarning is emitted in a module `#713 <https://github.com/pytest-dev/pytest-asyncio/issues/713>`_
+- Fixes various bugs that caused an internal pytest error during test collection `#711 <https://github.com/pytest-dev/pytest-asyncio/issues/711>`_ `#713 <https://github.com/pytest-dev/pytest-asyncio/issues/713>`_ `#719 <https://github.com/pytest-dev/pytest-asyncio/issues/719>`_
+
 
 
 0.23.2 (2023-12-04)

--- a/docs/source/reference/changelog.rst
+++ b/docs/source/reference/changelog.rst
@@ -6,6 +6,7 @@ Changelog
 ===================
 - Fixes a bug that caused event loops to be closed prematurely when using async generator fixtures with class scope or wider in a function-scoped test `#708 <https://github.com/pytest-dev/pytest-asyncio/issues/708>`_
 - Fixes a bug that caused an internal pytest error when using unittest.SkipTest in a module `#711 <https://github.com/pytest-dev/pytest-asyncio/issues/711>`_
+- Fixes a bug that caused an internal pytest error when an ImportWarning is emitted in a module `#713 <https://github.com/pytest-dev/pytest-asyncio/issues/713>`_
 
 
 0.23.2 (2023-12-04)

--- a/pytest_asyncio/plugin.py
+++ b/pytest_asyncio/plugin.py
@@ -26,10 +26,8 @@ from typing import (
     Union,
     overload,
 )
-from unittest import SkipTest
 
 import pytest
-from _pytest.outcomes import OutcomeException
 from pytest import (
     Class,
     Collector,
@@ -608,25 +606,28 @@ def pytest_collectstart(collector: pytest.Collector):
 
     # @pytest.fixture does not register the fixture anywhere, so pytest doesn't
     # know it exists. We work around this by attaching the fixture function to the
-    # collected Python class, where it will be picked up by pytest.Class.collect()
+    # collected Python object, where it will be picked up by pytest.Class.collect()
     # or pytest.Module.collect(), respectively
-    try:
-        pyobject = collector.obj
-        # If the collected module is a DoctestTextfile, collector.obj is None
-        if pyobject is None:
-            return
-        pyobject.__pytest_asyncio_scoped_event_loop = scoped_event_loop
-    except (OutcomeException, Collector.CollectError):
+    if type(collector) is Module:
         # Accessing Module.obj triggers a module import executing module-level
         # statements. A module-level pytest.skip statement raises the "Skipped"
         # OutcomeException or a Collector.CollectError, if the "allow_module_level"
         # kwargs is missing. These cases are handled correctly when they happen inside
         # Collector.collect(), but this hook runs before the actual collect call.
-        return
-    except SkipTest:
-        # Users may also have a unittest suite that they run with pytest.
-        # Therefore, we need to handle SkipTest to avoid breaking test collection.
-        return
+        # Therefore, we monkey patch Module.collect to add the scoped fixture to the
+        # module before it runs the actual collection.
+        def _patched_collect():
+            collector.obj.__pytest_asyncio_scoped_event_loop = scoped_event_loop
+            return collector.__original_collect()
+
+        collector.__original_collect = collector.collect
+        collector.collect = _patched_collect
+    else:
+        pyobject = collector.obj
+        # If the collected module is a DoctestTextfile, collector.obj is None
+        if pyobject is None:
+            return
+        pyobject.__pytest_asyncio_scoped_event_loop = scoped_event_loop
     # When collector is a package, collector.obj is the package's __init__.py.
     # pytest doesn't seem to collect fixtures in __init__.py.
     # Using parsefactories to collect fixtures in __init__.py their baseid will end

--- a/tests/test_import.py
+++ b/tests/test_import.py
@@ -1,0 +1,18 @@
+from textwrap import dedent
+
+from pytest import Pytester
+
+
+def test_import_warning(pytester: Pytester):
+    pytester.makepyfile(
+        dedent(
+            """\
+                raise ImportWarning()
+
+                async def test_errors_out():
+                    pass
+            """
+        )
+    )
+    result = pytester.runpytest("--asyncio-mode=auto")
+    result.assert_outcomes(errors=1)

